src/MIDebugEngine/Engine.Impl/ExceptionManager.cs

```diff
@@ -473,24 +510,61 @@ private async Task UpdateCatagory(Guid categoryId, ExceptionCategorySettings cat
                 if (grouping.Key == categorySettings.CategoryState)
                 {
                     // A request to set an exception to the same state as the category is redundant unless we have previously changed the state of that exception to something else
-                    exceptionNames = exceptionNames.Intersect(categorySettings.CurrentRules.Keys);
+                    lock (categorySettings.CurrentRules)
+                    {
+                        exceptionNames = exceptionNames.Intersect(categorySettings.CurrentRules.Keys);
+                    }
                     if (!exceptionNames.Any())
                     {
                         continue; // no exceptions left, so ignore this group
                     }
                 }
 
-                IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, exceptionNames, grouping.Key);
+                bool isBreakThrown = grouping.Key.HasFlag(ExceptionBreakpointStates.BreakThrown);
 
-                int count = exceptionNames.Zip(breakpointIds, (exceptionName, breakpointId) =>
+                if (!categorySettings.CategoryState.HasFlag(ExceptionBreakpointStates.BreakThrown) && isBreakThrown)
                 {
-                    categorySettings.CurrentRules[exceptionName] = breakpointId;
-                    return 1;
-                }).Sum();
+                    IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, exceptionNames, grouping.Key);
+
+                    lock (categorySettings.CurrentRules)
+                    {
+                        int count = exceptionNames.Zip(breakpointIds, (exceptionName, breakpointId) =>
+                        {
+                            // remove old breakpoint if exceptionName is in categorySettings.CurrentRules.Keys
+                            if (categorySettings.CurrentRules.ContainsKey(exceptionName))
+                            {
+                                _commandFactory.RemoveExceptionBreakpoint(categoryId, new ulong[] { categorySettings.CurrentRules[exceptionName] });
+                            }
+                            categorySettings.CurrentRules[exceptionName] = breakpointId;
+                            return 1;
+                        }).Sum();
 
 #if DEBUG
-                Debug.Assert(count == exceptionNames.Count());
+                        Debug.Assert(count == exceptionNames.Count());
 #endif
+                    }
+                }
+                else if (grouping.Key != categorySettings.CategoryState && !isBreakThrown)
+                {
+                    // Send warning when there are unchecked exceptions in a checked exceptions category
+                    HostOutputWindow.WriteLaunchError("There are unchecked exceptions in a checked exceptions category.");
```

Comment:
>HostOutputWindow.WriteLaunchError("There are unchecked exceptions in a checked exceptions category."); [](http://example.com/codeflow?start=20&length=102)

A few changes:

1. `HostOutputWindow.WriteLaunchError` is supposed to just be for launch errors, so you don't want that. Instead you want a new method on EngineCallback to send a message event with `enum_MESSAGETYPE.MT_OUTPUTSTRING`. Something like `_callback.OnOutputWindowWarning(string message)`
2. You want this string to be in a .resx so it can be localized.
3. I think I would suggest the text be tweaked to give the user more context. Something like:
`Warning: one or more C++ exceptions are unchecked while '<All C++ Exceptions not in this list>' is enabled. This is unsupported for GDB and the debugger will break on all C++ exceptions.\n'

Reply:
If I may ask, why is a new method on EngineCallback better than
```
_callback.OnOutputMessage(new OutputMessage("Warning: one or more C++ exceptions are unchecked while \'<All C++ Exceptions not in this list>\' is enabled. This is unsupported for GDB and the debugger will break on all C++ exceptions.\n", enum_MESSAGETYPE.MT_OUTPUTSTRING, OutputMessage.Severity.Warning));
```

Reply:
Oh, I missed that. That works.

Reply:
Okay. Will add the string to the .resx and then will commit my changes. Thanks, @gregg-miskelly!